internal/controller/postgrescluster/instance.go

```diff
@@ -1672,18 +1648,19 @@ func (r *Reconciler) observeUpgradeEnv(ctx context.Context,
 				Reason:             "PGUpgradeComplete",
 				Message:            "pg_upgrade completed successfully",
 			})
-			meta.SetStatusCondition(&cluster.Status.Conditions, metav1.Condition{
-				ObservedGeneration: cluster.GetGeneration(),
-				Type:               ConditionPostPGUpgrade,
-				Status:             metav1.ConditionTrue,
-				Reason:             "PGUpgradeComplete",
-				Message:            "pg_upgrade completed successfully, post upgrade tasks required.",
-			})
+			condition := meta.FindStatusCondition(cluster.Status.Conditions,
+				ConditionPGBackRestPostPGUpgrade)
+			if condition != nil {
+				condition.Status = metav1.ConditionTrue
+				meta.SetStatusCondition(&cluster.Status.Conditions, *condition)
+			}
 			// TODO: remove guard with move to controller-runtime 0.9.0 https://issue.k8s.io/99714
 			if len(cluster.Status.Conditions) > 0 {
 				meta.RemoveStatusCondition(&cluster.Status.Conditions,
 					ConditionPGUpgradeProgressing)
 			}
+			// Once the upgrade is complete, note the upgraded postgres version
+			cluster.Status.PostgresVersion = cluster.Spec.PostgresVersion
```

Comment:
I honestly struggled quite a bit with the proper place to put this.  My primary concern is knowing that when we write the PG version to the status, we 100% know the PG version we write is what is actually in the primary `PGDATA` volume.

My primary concern/issue with the prior approach (in which we update the PG version when populate the `SystemIdentifier`) is that if the PG version was updated in the spec while upgrading was disabled, we were simply writing the version from the spec into the status.

So what I started thinking about are the points we know 100% what version of PG is represented in the current primary `PGDATA` volume.  And ultimately I settled on the upgrade job, since when that `pg_upgrade` job completes, we can say for sure what version of PG in on disk (and I'm now wondering if I should be grabbing the version based on the PG version information annotated on the completed job, rather than the spec - since technically someone could change the PG version again in the spec while an upgrade is running).

Now what I don't like about this approach is that there is no way to repopulate the version if the upgrade job is deleted, and the PG version in the status has been removed.  Additionally, the PG version is never populated until an upgrade completes. However, by forcing the user to provide a "from version" we're at least eliminating some of the dependency on this status when determining whether or not to initiate an upgrade.

What would be great would be to have some way to detect the PG version on each reconcile, without having to exec into an instance each time.  E.g. it would be great if Patroni surfaced PG version information in some of the information it write to the various instance Pod (e.g. the annotation Patroni populates).  However, at the moment I have not yet identified a specific solution.

So for now, considering the upgrade functionality is the only thing that relies on the accuracy of the PG version in the status, I aligned it with that.  However, definitely open to any thoughts/suggestions.

Reply:
Thanks for the explanation. I also had some misgivings about the earlier implementation, and I agree that this is a more definitive approach. While it would be preferable to set this more often, I also can't think of a great way to do this currently. 

I do wonder if it would be helpful to add a note clarifying this behavior, perhaps where this status is defined in `pkg/apis/postgres-operator.crunchydata.com/v1beta1/postgrescluster_types.go`?

Reply:
I went ahead and update the description for the PG version in the `status`:

> Stores the current PostgreSQL major version.  This field is updated following a successful major PostgreSQL upgrade in order to track the PostgresCluster's PostgreSQL version across multiple major upgrades.